### PR TITLE
Finetuning benchmarking experiments

### PR DESCRIPTION
**Pull request by KeremTurgutlu**
This branch was used in running QLoRA benchmarking experiments using different GPU setups and configs. It also includes few LoRA experiments. 

- Added DDP as an option.
- Default to 128 samples in `alpaca_sample`.
- Added [sql dataset](https://huggingface.co/datasets/knowrohit07/know_sql) for training.
- Constant `lr_scheduler` is now a no-op to avoid the cpu-offloading overhead, which requires more investigation. In early experiments we've observed that having `lr_scheduler.step()` while using cpu-offloading slows down training by a factor of 1.5.
- Added wd to adamw optimizer.
- Refactor auto-wrap policy to its own function like other utilities.
- Added a custom qlora module (adapted from PEFT), in early benchmarking experiments it is observed to be more memory efficient compared to PEFT but a longer training run might be required to verify there isn't any bugs.
- Set SDPA as the attention implementation, with [torch 2.2 release](https://pytorch.org/blog/pytorch2-2/) it supports flash-attn 2 implementation and best kernel will be picked by default. This might not be that necessary since HF already uses it as the default for the known models but it is more of a comment.

```python
import torch
# Will throw kernel not available error when attn_mask is used.
# Optionally use the context manager to ensure one of the fused kernels is run
query = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
key = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
value = torch.rand(32, 8, 128, 64, dtype=torch.float16, device="cuda")
attn_mask = torch.ones(128, dtype=torch.bool, device="cuda")
with torch.backends.cuda.sdp_kernel(enable_flash=True, 
                                    enable_math=False, 
                                    enable_mem_efficient=False):
    torch.nn.functional.scaled_dot_product_attention(query,key,value,attn_mask=attn_mask)
```

- Turn gradient clipping off by default, similar to `lr_scheduler.step()` this introduces a significant slow down (3x-4x) when used with cpu-offloading, `llama-recipes` uses a scheduler which is only stepped at the end of each epoch. For fancier schedulers like cosine annealer that step with param updates, this needs more investigation.

### CODE REVIEW

**Comment by johnowhitaker** [comment]
Looking good! @warner-benjamin want to skim then merge?

File: train.py [hunk 1]
```diff
@@ -352,11 +368,72 @@ def get_optimizer(model:nn.Module, args:Dict):
...
+            PromptEmbedding,
+            transformer_layer_name,
+        ),
+    )
+    return functools.partial(_or_policy, policies=[lambda_policy, transformer_wrap_policy])
+
+
+# Custom QLORA module.
+class QLORA(nn.Module):
+    def __init__(self, base_layer, lora_rank, lora_alpha, lora_dropout, device="cpu"):
```
Comment (warner-benjamin) [inline train.py:402]: Would it be better to init the LoRA layers on the `base_layer.device` rather than CPU?

Comment (jph00) [reply train.py:402]: Does it matter -- they're really small anyway right?

Comment (KeremTurgutlu) [reply train.py]: That was also my thinking also, but made the change nevertheless.
